Fix rare crash when a transaction executes after its shared object assignments have been deleted

bingyanglin opened the pull request:
# Description of change

Fix rare crash when a transaction executes after its shared object assignments have been deleted.
Note that in the upstream changes there is a new `fatal!` macro, where will be addressed in another PR with issue #4756 

## Links to any relevant issues

fixes #4690

## Type of change

- Bug fix

## How the change has been tested

Ran the local network with `RUST_LOG=info cargo run --release --bin iota start --force-regenesis --with-faucet`

## Change checklist

Tick the boxes that are relevant to your changes, and delete any items that are not.

- [x] I have followed the contribution guidelines for this project
- [x] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have checked that new and existing unit tests pass locally with my changes

bingyanglin committed:
Apply upstream change of MystenLabs/sui@3aac32e

bingyanglin committed:
Fix errors

bingyanglin committed:
Fix typo

bingyanglin committed:
Fix error

bingyanglin committed:
Clippy

bingyanglin committed:
Merge branch 'develop' into core-node/fix/sui@3aac32e

DaughterOfMars commented on crates/iota-core/src/authority.rs:
```diff
@@ -1158,7 +1159,20 @@ impl AuthorityState {
         let tx_digest = certificate.digest();
-        let input_objects = self.read_objects_for_execution(certificate, epoch_store)?;
+
+        // prevent concurrent executions of the same tx.
```
```suggestion
        // Acquire a lock to prevent concurrent executions of the same transaction.
```

DaughterOfMars commented on crates/iota-core/src/authority.rs:
```diff
@@ -1158,7 +1159,20 @@ impl AuthorityState {
+
+        // prevent concurrent executions of the same tx.
+        let tx_guard = epoch_store.acquire_tx_guard(certificate).await?;
+        // The cert could have been processed by a concurrent attempt of the same cert,
+        // so check if the effects have already been written.
+        if let Some(effects) = self
```
```suggestion
        
        // The cert could have been processed by a concurrent attempt of the same cert,
        // so check if the effects have already been written.
        if let Some(effects) = self
```

DaughterOfMars commented on crates/iota-core/src/authority.rs:
```diff
@@ -1188,6 +1195,7 @@ impl AuthorityState {
 
     pub fn read_objects_for_execution(
         &self,
+        tx_lock: &CertLockGuard,
```
Some places this is `tx_lock`, others `tx_guard`. I would suggest sticking with one

bingyanglin replied:
Good suggestion, looks reasonable to change new added ones to be `tx_guard`.

bingyanglin replied:
So after double check now we call them `tx_lock` if it is `CertLockGuard`, and `tx_guard` if it is `GaugeGuard<'_>` or `CertTxGuard`, i.e., no `Lock` naming in the struct.
Then I think current naming might be clear already?

DaughterOfMars replied:
Alright

DaughterOfMars commented on crates/iota-core/src/authority/authority_per_epoch_store.rs:
```diff
@@ -131,12 +131,22 @@ pub(crate) type EncG = bls12381::G2Element;
 // to re-implement a retry / write-ahead-log at that point.
-pub struct CertLockGuard(#[expect(unused)] MutexGuard);
-pub struct CertTxGuard(#[expect(unused)] CertLockGuard);
+pub struct CertLockGuard(#[allow(unused)] MutexGuard);
+pub struct CertTxGuard(#[allow(unused)] CertLockGuard);
```
If this is no longer unused, then remove the tags altogether please

bingyanglin replied:
Modified at [a1dbc6f](https://github.com/iotaledger/iota/pull/4757/commits/a1dbc6f2329904210cd6520dfc0478f3eb13fd7d). Note that for the `MutexGuard` the tag is still needed.

DaughterOfMars replied:
Then it should be `expect`

bingyanglin replied:
Enhanced the tag at [a8d3303](https://github.com/iotaledger/iota/pull/4757/commits/a8d33036b54d7a794c485d5feb9d9490490eb371).

DaughterOfMars commented on crates/iota-core/src/authority/authority_per_epoch_store.rs:
```diff
@@ -131,12 +131,22 @@ pub(crate) type EncG = bls12381::G2Element;
+}
+
+impl CertLockGuard {
+    pub fn dummy_for_tests() -> Self {
```
Do we like `dummy`?

bingyanglin replied:
better suggesion?

DaughterOfMars replied:
`guard_for_tests` is probably fine?

bingyanglin replied:
Modified at [fa3d2b6](https://github.com/iotaledger/iota/pull/4757/commits/fa3d2b6563477d4c2c2274f9bea66ab40ebd7e0f).

DaughterOfMars commented on crates/iota-core/src/transaction_input_loader.rs:
```diff
@@ -134,10 +136,11 @@ impl TransactionInputLoader {
         &self,
         shared_lock_store: &impl GetSharedLocks,
         tx_key: &TransactionKey,
+        _tx_lock: &CertLockGuard, // see below for why this is needed
```
I don't see anything below?

bingyanglin replied:
The reason is the following, as included in this PR, and same as the MystenLabs/sui@3aac32e
```rust
                            // Important to hold the _tx_lock here - otherwise it would be possible
                            // for a concurrent execution of the same tx to enter this point after
                            // the first execution has finished and the
                            // shared locks have been deleted.
```

DaughterOfMars replied:
I would move this comment up to the param site.

bingyanglin replied:
Modified at [93f423b](https://github.com/iotaledger/iota/pull/4757/commits/93f423b897183cc74dc64b0552b8c58e860dc7cb).

bingyanglin committed:
Update crates/iota-core/src/authority.rs

Co-authored-by: DaughterOfMars <chloedaughterofmars@gmail.com>

bingyanglin committed:
Update crates/iota-core/src/authority.rs

Co-authored-by: DaughterOfMars <chloedaughterofmars@gmail.com>

Thoralf-M commented:
Could you please choose another PR title, which makes it clear what this change is about? `upstream change` isn't very helpful

DaughterOfMars commented on crates/iota-core/src/transaction_manager.rs:
```diff
@@ -438,15 +438,34 @@ impl TransactionManager {
+                            {
+                                return None;
+                            }
+                            panic!("Failed to get input object keys: {:?}", e);
```
Why are we using panics instead of propagating errors?

bingyanglin replied:
I think this should not happen?

DaughterOfMars replied:
I guess this is just how this code works 🤷‍♀️ as there seems to be no other error propagation

DaughterOfMars commented on crates/iota-core/src/unit_tests/authority_tests.rs:
```diff
@@ -4713,6 +4713,7 @@ async fn test_shared_object_transaction_ok() {
         .epoch_store_for_testing()
         .get_shared_locks(&certificate.key())
         .expect("Reading shared locks should not fail")
+        .expect("Locks should be set")
```
These comments are in the wrong format. They should describe the error, not what should occur. More below and also some are incorrectly capitalized

bingyanglin replied:
Sorry I am not following what format is wrong? Can you give me an example?

DaughterOfMars replied:
1. These messages should be all lowercase
2. The phrasing should be "locks are not set" instead of "locks should be set", I.E. Descriptive not prescriptive

bingyanglin replied:
Fixed at [d206cd8](https://github.com/iotaledger/iota/pull/4757/commits/d206cd846c0a10ff1496e207cf2a504492d43e66).

bingyanglin commented:
> Could you please choose another PR title, which makes it clear what this change is about? `upstream change` isn't very helpful

Good suggestion. Changed the title, and please feel free to suggest a better naming. Thanks.

DaughterOfMars commented on crates/iota-single-node-benchmark/src/mock_storage.rs:
```diff
@@ -57,7 +57,7 @@ impl InMemoryObjectStore {
         input_object_kinds: &[InputObjectKind],
     ) -> IotaResult<InputObjects> {
-        let shared_locks_cell: OnceCell<HashMap<_, _>> = OnceCell::new();
+        let shared_locks_cell: OnceCell<Option<HashMap<_, _>>> = OnceCell::new();
```
Can you explain this change?

bingyanglin replied:
Yes. As the `shared_locks` was also changed in the following code.

DaughterOfMars replied:
But why?

bingyanglin replied:
I think it's for better handling when the `shared_locks` returns `None`. When it is `None`, it will be handled by the following error
```rust 
IotaError::GenericAuthority {
    error: "Shared object locks should have been set.".to_string(),
}
```

DaughterOfMars replied:
Yes, but it is set by `get_or_init`. This whole change seems very suspicious, as it adds an `Option` only so that it can fail.

bingyanglin replied:
Tbh I don't have other clues, but I would prefer use this change as we can handle the `None` case better.

bingyanglin committed:
Fmt

thibault-martinez commented:
> Good suggestion. Changed the title, and please feel free to suggest a better naming. Thanks.

IMO I would rather have the name be explicit about what it is fixing than just saying it's porting a commit

bingyanglin commented:
> > Good suggestion. Changed the title, and please feel free to suggest a better naming. Thanks.
> 
> IMO I would rather have the name be explicit about what it is fixing than just saying it's porting a commit

Agreed. Changed the title to be explicit and the same as the ported commits, so we can easier map them in the future.

bingyanglin committed:
Remove unused macro for CertLockGuard

bingyanglin committed:
Move the _tx_lock comments

bingyanglin committed:
Rename dummy_for_tests to guard_for_tests

bingyanglin committed:
Refine expect mesage

bingyanglin committed:
Merge branch 'develop' into core-node/fix/sui@3aac32e

bingyanglin committed:
Merge branch 'develop' into core-node/fix/sui@3aac32e

muXxer commented on crates/iota-core/src/transaction_input_loader.rs:
```diff
@@ -161,17 +167,20 @@ impl TransactionInputLoader {
+                        .as_ref()
+                        .unwrap_or_else(|| {
+                            // _tx_lock is held, so this should not happen
+                            panic!("Failed to get shared locks for transaction {tx_key:?}");
```
Should we add a TODO that "fatal" should be used here in the future?

bingyanglin replied:
Thanks! Fixed at [bec3e9b](https://github.com/iotaledger/iota/pull/4757/commits/bec3e9b139251a8fb5a5d6595c681c517ae9030d).

alexsporn replied:
@bingyanglin `fatal!` shoud be available now, so please update

muXxer commented on crates/iota-core/src/transaction_manager.rs:
```diff
@@ -438,15 +438,34 @@ impl TransactionManager {
         let certs: Vec<_> = certs
             .into_iter()
-            .map(|(cert, fx_digest)| {
+            .filter_map(|(cert, fx_digest)| {
```
Maybe add a TODO in line 423 that we need to add fatal there.

bingyanglin replied:
Thanks for pointing this out! Fixed at [bec3e9b](https://github.com/iotaledger/iota/pull/4757/commits/bec3e9b139251a8fb5a5d6595c681c517ae9030d).

muXxer commented on crates/iota-core/src/transaction_manager.rs:
```diff
@@ -438,15 +438,34 @@ impl TransactionManager {
+                        Ok(keys) => keys,
+                        Err(e) => {
+                            // Because we do not hold the transaction lock during enqueue, it is
+                            // possible that the transaction was
```
maybe format the comment a bit nicer. with our changed line length settings this looks weird.

bingyanglin replied:
Agreed! Missed this. Fixed at [cdeb224](https://github.com/iotaledger/iota/pull/4757/commits/cdeb22415fc93c9e36e1d1b009bf2c7622cb5959).

muXxer commented on crates/iota-core/src/unit_tests/authority_tests.rs:
```diff
@@ -2069,15 +2069,15 @@ async fn test_conflicting_transactions() {
             gas_info
                 .lock_for_debugging
-                .expect("gas should be locked")
+                .expect("gas is not locked")
```
where did that changed? :D

bingyanglin replied:
So basically this is suggested by @DaughterOfMars [here](https://github.com/iotaledger/iota/pull/4757#discussion_r1913411569) where
```bash
1. These messages should be all lowercase
2. The phrasing should be "locks are not set" instead of "locks should be set", I.E. Descriptive not prescriptive
```

bingyanglin committed:
Add TODO comments for fatal! usage

bingyanglin committed:
Modify the comments

bingyanglin committed:
Enhance tag

bingyanglin committed:
Merge branch 'develop' into core-node/fix/sui@3aac32e

bingyanglin committed:
Merge branch 'develop' into core-node/fix/sui@3aac32e

alexsporn commented on crates/iota-core/src/transaction_manager.rs:
```diff
@@ -420,6 +420,7 @@ impl TransactionManager {
                     .transaction_cache_read
                     .is_tx_already_executed(&digest)
                     .unwrap_or_else(|err| {
+                        // TODO: use fatal! once it is available
```
Same here

alexsporn commented on crates/iota-core/src/transaction_manager.rs:
```diff
@@ -438,15 +439,34 @@ impl TransactionManager {
+                                return None;
+                            }
+                            // TODO: use fatal! once it is available
+                            panic!("Failed to get input object keys: {:?}", e);
```
Same here

bylin-mh replied:
Thanks! Modified at [db7870d](https://github.com/iotaledger/iota/pull/4757/commits/db7870d37f7dc3acc6baae9d840e591447509887).

alexsporn reviewed (changes requested):
Please update with `develop` and fix the introduced TODOs

bingyanglin committed:
Merge branch 'develop' into core-node/fix/sui@3aac32e

bingyanglin committed:
Use fatal macro
